apps/hash-frontend/src/blocks/page/mention-view/mention-display.tsx

```diff
@@ -1,156 +1,328 @@
-        if (!users || (usersLoading && !users.length)) {
-          /** @todo - What should the href be here? */
-          return {
-            title: "User",
-            href: userHref,
-            icon: "@",
-          };
-        }
+  const { entityId } = mention;
+  const { entitySubgraph, loading } = useEntityById(entityId);
+  const contentRef = useRef<HTMLDivElement>(null);
 
-        // Once the query loads, either display the found name, or display "Unknown User" if the user doesn't exist in the users array
-        const matchingUser = users.find(
-          (user) => user.entityRecordId.entityId === entityId,
-        );
-
-        if (matchingUser) {
-          return {
-            title: matchingUser.preferredName,
-            href: userHref,
-            icon: "@",
-          };
-        }
+  const [popoverOpen, setPopoverOpen] = useState(false);
 
-        /** @todo - What should the href be here? */
-        return {
-          title: "Unknown User",
-          href: `#`,
-          icon: "@",
-        };
-      }
+  const entity = useMemo(
+    () => (entitySubgraph ? getRoots(entitySubgraph)[0] : undefined),
+    [entitySubgraph],
+  );
 
-      case "page": {
-        const page = pages.find(
-          (potentialPage) =>
-            potentialPage.metadata.recordId.entityId === entityId,
-        );
+  const entityLabel = useMemo(
+    () => (entitySubgraph ? generateEntityLabel(entitySubgraph) : undefined),
+    [entitySubgraph],
+  );
 
-        let pageTitle = "";
+  const getOwnerForEntity = useGetOwnerForEntity();
 
-        // Only set the title to "Page" if the query hasn't returned yet
-        if (pagesLoading && !pages.length) {
-          pageTitle = "Page";
-        } else {
-          // Once the query loads, either display the found title, or display "Unknown Page" if the page doesn't exist in the page array
-          pageTitle = page?.title ?? "Unknown Page";
-        }
+  const entityOwnerShortname = useMemo(() => {
+    if (entity) {
+      const { shortname } = getOwnerForEntity(entity);
 
+      return shortname;
+    }
+  }, [entity, getOwnerForEntity]);
+
+  const entityHref = useMemo(() => {
+    if (entity) {
+      return `/@${entityOwnerShortname}/entities/${extractEntityUuidFromEntityId(
+        entity.metadata.recordId.entityId,
+      )}`;
+    }
+  }, [entity, entityOwnerShortname]);
+
+  const title = useMemo(() => {
+    if (!entitySubgraph) {
+      return undefined;
+    }
+
+    if (
+      mention.kind === "entity" ||
+      mention.kind === "page" ||
+      mention.kind === "user"
+    ) {
+      return entityLabel;
+    } else if (mention.kind === "outgoing-link") {
+      const outgoingLinkAndTargetEntities = getOutgoingLinkAndTargetEntities(
+        entitySubgraph,
+        entityId,
+      ).find(
+        ({ linkEntity: linkEntityRevisions }) =>
+          linkEntityRevisions[0]?.metadata.recordId.entityId ===
+          mention.linkEntityId,
+      );
```

Comment:
If a link is replaced, there will be a new link entity created, rather than the existing one updated. So this approach will only work so long as the link isn't replaced (I'm guessing that the chosen link will not be found in the subgraph once it's archived, or – if it is in the subgraph – it will be out of date).

An approach which would be resilient to 
(a) the link being replaced, and 
(b) the link type being updated 
would be to store the link type's base URL instead, and then find the link from the entity of that type.

The downside to this would be where there are multiple links of the same type. For now we could just take the first one.

Reply:
That's a good point, I've applied this suggestion in https://github.com/hashintel/hash/pull/3244/commits/0ded7492a47c3c492730569cc7da49bfc55d607f